django/cantusdb_project/main_app/tests/test_views.py

```diff
@@ -749,6 +749,237 @@ def test_keyword_search_contains(self):
         )
         self.assertIn(chant, response.context["chants"])
 
+    def test_source_link_column(self):
+        siglum = "Sigl-01"
+        source = make_fake_source(published=True, siglum=siglum)
+        url = source.get_absolute_url()
+        fulltext = "manuscript full text"
+        search_term = "full"
+        chant = make_fake_chant(
+            source=source,
+            manuscript_full_text_std_spelling=fulltext
+        )
+        response = self.client.get(
+            reverse("chant-search"), {"keyword": search_term, "op": "contains"}
+        )
+        html = str(response.content)
+        self.assertIn(siglum, html)
```

Comment:
Yes, they would be.

In writing these tests, I've been following a pattern of testing for one or a few specific details, and then testing that they all come together correctly. Though I hadn't fully thought this out when I was writing them, I think the reason I did it is to communicate intent to future developers. The first two `assertIn`s are my way of saying, "yes, we really do expect that both the siglum and the URL are both present and correct", while the third one is just there to make sure everything is fitting together properly. I'm thinking of these tests, I suppose, as indications/hints about what the important elements are in each of our views and functions (in addition, of course, to the tests being tools to verify that nothing is broken).

Another way of putting it: if, in the future, someone tweaks the relevant <a> tag in the template, I expect them to adjust the final `assert` to reflect the change, but I also expect them to make sure they know what they're doing before they delete either of the first two `assert`s.

Do you think this approach makes sense?

Reply:
Thanks for explaining! This is a thoughtful approach...and you've consistently applied it, so I think it works. I hadn't thought about tests as (among other things) an additional form of documentation ~ I like that concept.

Reply:
A maybe-related question I have is the reasoning behind creating minimal fake objects for each test. For example, to test the feast and office columns, you create a fake chant with a fake feast and a fake chant with a fake office, respectively, but minimal other attributes. It seems like another approach would be to create a fake chant with both a fake feast and fake office and test both columns against that one chant, testing all those features with multiple assert statements. Are you trying to stick with the minimum object that will pass the test in each case so that that's clear to other's looking at the code (ie. a  similar "documentary" purpose" as above)?

I'm mostly asking because in reviewing this I realize I have no idea what's going on with Cantus Ultimus's tests//haven't touched them since I joined the lab (I think maybe Eli did something with them early in summer 2022?) so I'm now curious about the way you're thinking about them and think it might be good for me to get some sense of what's going on over at CU.

Reply:
This is a good point, and one I haven't thought about very deeply! One thing that comes to mind - not sure if this is exactly what you're thinking of - is that if we test, say, filtering of search results by multiple fields at the same time, the number of tests we write becomes an exercise in combinatorics.

In any case, it would probably be worth having a couple of tests at the bottom of each test class that build one or several objects with as many attributes as we expect to see "in the real world" for which we test that things come together properly, without aiming to be exhaustive.

Reply:
> I realize I have no idea what's going on with Cantus Ultimus's tests//haven't touched them since I joined the lab

Spending some time to make sure the test suite is up-to-date and covers everything important is something I can hardly recommend highly enough! For me, not only do I find that good tests alleviate most of my concerns that I've introduced a subtle bug someplace whenever I make a change, but I also find that writing tests can clarify what a new function/feature will do.

If you need any help with this, or have any questions, let me know: I enjoy writing tests, and would be happy to help wherever I can!